gh-138122: Replace --interval with --sampling-rate

lkollar opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->

Towards #142927.

CC: @pablogsal @ivonastojanovic 

<!-- gh-issue-number: gh-138122 -->
* Issue: gh-138122
<!-- /gh-issue-number -->

lkollar committed:
Standardise time units

pablogsal commented on Doc/library/profiling.sampling.rst:
```diff
@@ -296,8 +296,8 @@ The default configuration works well for most use cases:
-   * - Default for ``--interval`` / ``-i``
-     - 100 µs between samples (~10,000 samples/sec)
+   * - Default for ``--sampling-rate`` / ``-r``
+     - 10 kHz
```
Please update the documentation to reflect the actual 1 kHz default

pablogsal commented on Lib/profiling/sampling/cli.py:
```diff
@@ -286,16 +288,51 @@ def _run_with_sync(original_cmd, suppress_output=False):
-        help="sampling interval",
+        "-r",
+        "--sampling-rate",
+        type=_parse_sampling_rate,
```
Nit: The argument is named `sampling_rate` but after parsing it stores the *interval* in microseconds, not the rate in Hz. This is kind of confusing in the rest of the code

pablogsal commented on Lib/profiling/sampling/cli.py:
```diff
@@ -286,16 +288,51 @@ def _run_with_sync(original_cmd, suppress_output=False):
+
+    match = _RATE_PATTERN.match(rate_str)
+    if not match:
+        raise argparse.ArgumentTypeError(
```
Nit: Let's add a hint about spaces in the error message since "10 khz" (with space) is rejected but users might try it

pablogsal commented on Lib/profiling/sampling/cli.py:
```diff
@@ -286,16 +288,51 @@ def _run_with_sync(original_cmd, suppress_output=False):
         return process
 
 
+_RATE_PATTERN = re.compile(r'^(\d+(?:\.\d+)?)(hz|khz|k)?$', re.IGNORECASE)
```
```suggestion
_RATE_PATTERN = re.compile(r'''
      ^                   # Start of string
      (                   # Group 1: The numeric value
          \d+             #   One or more digits (integer part)
          (?:\.\d+)?      #   Optional: decimal point followed by digits
      )                   #   Examples: "10", "0.5", "100.25"
      (                   # Group 2: Optional unit suffix
          hz              #   "hz" - hertz
          | khz           #   "khz" - kilohertz
          | k             #   "k" - shorthand for kilohertz
      )?                  #   Suffix is optional (bare number = Hz)
      $                   # End of string
  ''', re.VERBOSE | re.IGNORECASE)
```

pablogsal replied:
Haha you know how much I like to comment these :)

pablogsal reviewed (commented):
LGTM once the documentation is updated! 🚀

lkollar committed:
Replace --interval with --sampling-rate

Sampling rate is more intuitive to the number of samples per second
taken, rather than the intervals between samples.

pablogsal reviewed (approved):
LGTM! Great work 👌
